core/templates/components/ck-editor-helpers/ck-editor-copy-content-service.ts

```diff
@@ -0,0 +1,173 @@
+
+interface CkEditorCopyEvent {
+  rootElement: HTMLElement;
+  containedWidgetTagName?: string;
+}
+
+@Injectable({
+  providedIn: 'root'
+})
+export class CkEditorCopyContentService {
+  private copyEventEmitter = new EventEmitter<CkEditorCopyEvent>();
+  private ckEditorIdToSubscription: {[id: string]: Subscription} = {};
+
+  copyModeActive = false;
+
+  constructor(private htmlEscaperService: HtmlEscaperService) {}
+
+  /**
+   * Traverses up and down element ancestors/descendants, searching for widget
+   * tags.
+   * @param {HTMLElement} target The element target that contains or is a
+   *  descendant of a widget, or a plain HTML element/group.
+   * @returns {CkEditorCopyEvent} Returns an object of type CkEditorCopyEvent,
+   *  where rootElement is the root ancestor of target and is
+   *  always a child of angular-html-bind. containedWidgetTagName is the tag
+   *  name of the widget if found in ancestors or descendants.
+   */
+  private _handleCopy(target: HTMLElement): CkEditorCopyEvent {
+    let containedWidgetTagName;
+    let currentElement = target;
+
+    while (true) {
+      const currentTagName = currentElement.tagName.toLowerCase();
+      if (currentTagName.includes('-noninteractive-')) {
+        containedWidgetTagName = currentTagName;
+      }
+
+      if (currentElement.parentElement.tagName === 'ANGULAR-HTML-BIND') {
+        break;
+      }
+
+      currentElement = currentElement.parentElement;
+    }
+
+    let descendents = Array.from(target.childNodes);
+    while (descendents.length !== 0) {
+      let currentDescendent = descendents.shift();
+      const currentTagName = currentDescendent.nodeName.toLowerCase();
+      if (currentTagName.includes('-noninteractive-')) {
+        containedWidgetTagName = currentTagName;
+        break;
+      }
+
+      descendents = [
+        ...descendents,
+        ...Array.from(currentDescendent.childNodes)
+      ];
+    }
+
+    return { rootElement: currentElement, containedWidgetTagName };
+  }
+
+  /**
+   * Detects if element is to be copied as a widget or as a HTML element, and
+   * inserts into editor.
+   * @param {CKEDITOR.editor} editor The editor in which to create widget or
+   *  insert HTML.
+   * @param {HTMLElement} element The element to be copied.
+   * @param {string|undefined} containedWidgetTagName The name of the widget
+   *  in which element contains, if present.
+   */
+  private _handlePaste(
+      editor: CKEDITOR.editor | Partial<CKEDITOR.editor>,
+      element: HTMLElement,
+      containedWidgetTagName?: string
+  ) {
+    let elementTagName = (
+      containedWidgetTagName || element.tagName.toLowerCase());
+    let html = element.outerHTML;
+
+    if (!containedWidgetTagName) {
+      editor.insertHtml(html);
+    } else {
+      const widgetName = elementTagName.replace('-noninteractive-', '');
+
+      // Look for x-with-value="y" to extract and y.
+      //  (\w+): Any word containing [a-zA-Z0-9_] characters.
```

Comment:
Add info about this will match the widget name etc.

Reply:
Done